File: src/C3.Catalogue/Brands/BrandService.vb
Namespace Brands

    Public NotInheritable Class BrandService

        Private ReadOnly _repository As IBrandRepository

        Public Sub New(repository As IBrandRepository)
            If repository Is Nothing Then
                Throw New ArgumentNullException("repository")
            End If
            _repository = repository
        End Sub

        Public Function GetAll(notesFilter As String) As IList(Of Brand)
            Dim values As IEnumerable(Of Brand) = _repository.GetAll()
            If Not String.IsNullOrWhiteSpace(notesFilter) Then
                Dim filter As String = notesFilter.Trim()
                values = values.Where(
                    Function(value As Brand) (If(value.Notes, String.Empty)).IndexOf(
                        filter,
                        StringComparison.CurrentCultureIgnoreCase) >= 0)
            End If
            Return values.OrderBy(Function(value As Brand) value.Code, StringComparer.OrdinalIgnoreCase).ToList()
        End Function

        Public Function Find(code As String) As Brand
            Return _repository.FindByCode(NormalizeCode(code))
        End Function

        Public Function Create(draft As BrandDraft, addedAt As DateTime) As BrandOperationResult
            Dim validation As BrandOperationResult = ValidateDraft(draft)
            If Not validation.IsSuccess Then
                Return validation
            End If

            Dim code As String = NormalizeCode(draft.Code)
            If _repository.IsCodeInUse(code) Then
                Return BrandOperationResult.Failed(
                    BrandFailure.DuplicateCode,
                    "Brand code '" & code & "' is already in use.")
            End If

            Dim value As New Brand(draft.Name.Trim(), code, addedAt, NormalizeNotes(draft.Notes))
            Try
                _repository.Add(value)
                Return BrandOperationResult.Success(value)
            Catch ex As Exception
                Return BrandOperationResult.Failed(BrandFailure.StorageFailure, ex.Message)
            End Try
        End Function

        Public Function Update(code As String, draft As BrandDraft) As BrandOperationResult
            Dim normalizedCode As String = NormalizeCode(code)
            Dim existing As Brand = _repository.FindByCode(normalizedCode)
            If existing Is Nothing Then
                Return BrandOperationResult.Failed(BrandFailure.NotFound, "The selected brand no longer exists.")
            End If

            Dim validation As BrandOperationResult = ValidateName(draft)
            If Not validation.IsSuccess Then
                Return validation
            End If

            Dim updated As New Brand(
                draft.Name.Trim(),
                normalizedCode,
                existing.AddedAt,
                NormalizeNotes(draft.Notes))
            Try
                _repository.Update(updated)
                Return BrandOperationResult.Success(updated)
            Catch ex As Exception
                Return BrandOperationResult.Failed(BrandFailure.StorageFailure, ex.Message)
            End Try
        End Function

        Public Function Delete(code As String) As BrandOperationResult
            Dim normalizedCode As String = NormalizeCode(code)
            Dim existing As Brand = _repository.FindByCode(normalizedCode)
            If existing Is Nothing Then
                Return BrandOperationResult.Failed(BrandFailure.NotFound, "The selected brand no longer exists.")
            End If
            If _repository.IsReferencedByModel(normalizedCode) Then
                Return BrandOperationResult.Failed(
                    BrandFailure.ReferencedByModel,
                    "Brand '" & normalizedCode & "' cannot be deleted while cassette models use it.")
            End If

            Try
                _repository.Delete(normalizedCode)
                Return BrandOperationResult.Success(existing)
            Catch ex As Exception
                Return BrandOperationResult.Failed(BrandFailure.StorageFailure, ex.Message)
            End Try
        End Function

        Private Shared Function ValidateDraft(draft As BrandDraft) As BrandOperationResult
            Dim nameValidation As BrandOperationResult = ValidateName(draft)
            If Not nameValidation.IsSuccess Then
                Return nameValidation
            End If

            Dim code As String = NormalizeCode(draft.Code)
            If code.Length <> 2 OrElse Not code.All(Function(character As Char) character >= "A"c AndAlso character <= "Z"c) Then
                Return BrandOperationResult.Failed(
                    BrandFailure.InvalidCode,
                    "Brand code must contain exactly two letters (A-Z).")
            End If

            Return BrandOperationResult.Success(Nothing)
        End Function

        Private Shared Function ValidateName(draft As BrandDraft) As BrandOperationResult
            If draft Is Nothing OrElse String.IsNullOrWhiteSpace(draft.Name) Then
                Return BrandOperationResult.Failed(BrandFailure.NameRequired, "Brand name is required.")
            End If
            Return BrandOperationResult.Success(Nothing)
        End Function

        Private Shared Function NormalizeCode(value As String) As String
            Return If(value, String.Empty).Trim().ToUpperInvariant()
        End Function

        Private Shared Function NormalizeNotes(value As String) As String
            Return If(value, String.Empty)
        End Function

    End Class

End Namespace

File: src/C3.Catalogue/Brands/BrandService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace C3.Catalogue.Brands
{
    public sealed class BrandService
    {
        private readonly IBrandRepository repository;

        public BrandService(IBrandRepository repository)
        {
            if (repository == null)
            {
                throw new ArgumentNullException(nameof(repository));
            }

            this.repository = repository;
        }

        public IList<Brand> GetAll(string notesFilter)
        {
            IEnumerable<Brand> values = repository.GetAll();
            if (!string.IsNullOrWhiteSpace(notesFilter))
            {
                var filter = notesFilter.Trim();
                values = values.Where(value => (value.Notes ?? string.Empty).IndexOf(
                    filter,
                    StringComparison.CurrentCultureIgnoreCase) >= 0);
            }

            return values
                .OrderBy(value => value.Code, StringComparer.OrdinalIgnoreCase)
                .ToList();
        }

        public Brand Find(string code)
        {
            return repository.FindByCode(NormalizeCode(code));
        }

        public BrandOperationResult Create(BrandDraft draft, DateTime addedAt)
        {
            var validation = ValidateDraft(draft);
            if (!validation.IsSuccess)
            {
                return validation;
            }

            var code = NormalizeCode(draft.Code);
            if (repository.IsCodeInUse(code))
            {
                return BrandOperationResult.Failed(
                    BrandFailure.DuplicateCode,
                    "Brand code '" + code + "' is already in use.");
            }

            var value = new Brand(
                draft.Name.Trim(),
                code,
                addedAt,
                NormalizeNotes(draft.Notes));
            try
            {
                repository.Add(value);
                return BrandOperationResult.Success(value);
            }
            catch (Exception exception)
            {
                return BrandOperationResult.Failed(
                    BrandFailure.StorageFailure,
                    exception.Message);
            }
        }

        public BrandOperationResult Update(string code, BrandDraft draft)
        {
            var normalizedCode = NormalizeCode(code);
            var existing = repository.FindByCode(normalizedCode);
            if (existing == null)
            {
                return BrandOperationResult.Failed(
                    BrandFailure.NotFound,
                    "The selected brand no longer exists.");
            }

            var validation = ValidateName(draft);
            if (!validation.IsSuccess)
            {
                return validation;
            }

            var updated = new Brand(
                draft.Name.Trim(),
                normalizedCode,
                existing.AddedAt,
                NormalizeNotes(draft.Notes));
            try
            {
                repository.Update(updated);
                return BrandOperationResult.Success(updated);
            }
            catch (Exception exception)
            {
                return BrandOperationResult.Failed(
                    BrandFailure.StorageFailure,
                    exception.Message);
            }
        }

        public BrandOperationResult Delete(string code)
        {
            var normalizedCode = NormalizeCode(code);
            var existing = repository.FindByCode(normalizedCode);
            if (existing == null)
            {
                return BrandOperationResult.Failed(
                    BrandFailure.NotFound,
                    "The selected brand no longer exists.");
            }

            if (repository.IsReferencedByModel(normalizedCode))
            {
                return BrandOperationResult.Failed(
                    BrandFailure.ReferencedByModel,
                    "Brand '" + normalizedCode +
                    "' cannot be deleted while cassette models use it.");
            }

            try
            {
                repository.Delete(normalizedCode);
                return BrandOperationResult.Success(existing);
            }
            catch (Exception exception)
            {
                return BrandOperationResult.Failed(
                    BrandFailure.StorageFailure,
                    exception.Message);
            }
        }

        private static BrandOperationResult ValidateDraft(BrandDraft draft)
        {
            var nameValidation = ValidateName(draft);
            if (!nameValidation.IsSuccess)
            {
                return nameValidation;
            }

            var code = NormalizeCode(draft.Code);
            if (code.Length != 2 || code.Any(character => character < 'A' || character > 'Z'))
            {
                return BrandOperationResult.Failed(
                    BrandFailure.InvalidCode,
                    "Brand code must contain exactly two letters (A-Z).");
            }

            return BrandOperationResult.Success(null);
        }

        private static BrandOperationResult ValidateName(BrandDraft draft)
        {
            if (draft == null || string.IsNullOrWhiteSpace(draft.Name))
            {
                return BrandOperationResult.Failed(
                    BrandFailure.NameRequired,
                    "Brand name is required.");
            }

            return BrandOperationResult.Success(null);
        }

        private static string NormalizeCode(string value)
        {
            return (value ?? string.Empty).Trim().ToUpperInvariant();
        }

        private static string NormalizeNotes(string value)
        {
            return value ?? string.Empty;
        }
    }
}
